core-server/src/main/java/org/glassfish/jersey/server/internal/routing/PathMatchingRouter.java

```diff
@@ -80,11 +92,42 @@
             }
         }
 
+        if (result == null && acceptedRouteCandidate != null) {
+            //method designator mismatched, but still go the route to get the proper status code
+            result = matchPathSelected(context, acceptedRouteCandidate, matchResultCandidate, tracingLogger);
+        }
+
         if (result == null) {
             // No match
             return Router.Continuation.of(context);
         }
 
         return result;
     }
+
+    private Router.Continuation matchPathSelected(final RequestProcessingContext context, final Route acceptedRoute,
+                                                  final MatchResult matchResult, final TracingLogger tracingLogger) {
+        // Push match result information and rest of path to match
+        context.routingContext().pushMatchResult(matchResult);
+        final Router.Continuation result = Router.Continuation.of(context, acceptedRoute.next());
+
+        // tracing
+        tracingLogger.log(ServerTraceEvent.MATCH_PATH_SELECTED, acceptedRoute.routingPattern().getRegex());
+
+        return result;
+    }
+
+    /**
+     * Return true iff the method is a sub-resource and not a locator and http method designator does not match
+     * the request http method designator
+     * @param route current route representing resource method / locator
+     * @param context Contains Request to check the http method
+     * @return false if method designator matches
+     */
+    private boolean designatorMismatch(final Route route, final RequestProcessingContext context) {
+        final String httpMethod = context.request().getMethod();
```

Comment:
I would negate the method, so the if statement does not have to use `!`.
Also we can be more "clever" with the `HEAD` method - e.g. first try it for `HEAD` and then for `GET`, as `GET` can also respond to `HEAD` requests.